### PR TITLE
50707 - Use mutable dictionaries in CDTDocumentRevision

### PR DESCRIPTION
**Pull request by mikerhodes**
_What_

Change CDTDocumentRevision's `body` and `attachment` properties into mutable dictionaries.

These dictionaries are wrapped with observer classes which report changes to the dictionaries and their deep contents, which allows us to report whether a particular document revision object has been changed since it was read from the database. This allows us to fix the conflict resolver tests, so they are re-enabled.

_Why_

The aim is to simplify the modification of document revisions for developers using the library. The PR allows developers to again modify the body and attachments, as they were able to on `CDTMutableDocumentRevision` objects.

So the previous way was this:

```objc
CDTDocumentRevision *rev = [datastore getDocumentWithId:id error:nil];
CDTMutableDocumentRevision *mutable = [rev mutableCopy];
mutable.body[@"foo"] = @"bar";
[datastore updateDocumentWithRevision:mutable error:nil];
```
with this PR, this is reduced to:
```objc
CDTDocumentRevision *rev = [datastore getDocumentWithId:id error:nil];
rev.body[@"foo"] = @"bar";
[datastore updateDocumentWithRevision:mutable error:nil];
```

_How_

First, the `body` and `attachment` properties are changed to be `NSMutableDictionary` objects. Internally, these are actually wrapped in NSMutableDictionary subclasses to enable the revision to observe changes as the users alters the body or attachments of the revision.

`isChanged` is a new property on `CDTDocumentRevision` and the new classes which wrap dictionaries and arrays to observe changes, `CDTChangedDictionary` and `CDTChangedArray` respectively. 

The class names  `CDTChangedDictionary` and `CDTChangedArray` seem a bit imperfect, but after thinking for a while I couldn't think of something better -- suggestions welcome.

A `CDTDocumentRevision` wraps the body and attachments passed into the `init...` method using a class method on `CDTChangedDictionary` which deep copies the contents of any dictionaries and attachments into new mutable dictionaries -- this helps prevent changes happening to the underlying containers which wouldn't be observed by the wrappers.

The `Changed*` classes report their changed status via both a property and a delegate. Parent containers register themselves as delegate on their children. The document revision object registers itself as delegate on the top level body and attachments dictionary. This way, changes to nested containers will cascade up to the document revision.

_Suggestions_

I think the easiest way to follow this PR -- if you'd rather I can break it up -- is:

1. CDTChangedArray -- this implements the primitive methods of NS(Mutable)Array, and fires off a message to its delegate.
1. CDTChangedDictionary -- similarly implements the primitive methods of NS(Mutable)Dictionary.
1. CDTChangedDictionary -- check the class methods for deep-wrapping dictionaries.
1. CDTDocumentRevision -- check the logic of isChanged, particularly whether it's always guaranteed to be set correctly on any change to body/attachments. The main hole here is non-wrapped objects, but I think for JSON, we only have primitive types which can't change without being replaced within the dict/array structure. I could be wrong.

For info on _primitive methods_, see https://mikeash.com/pyblog/friday-qa-2010-03-12-subclassing-class-clusters.html.

reviewer @rhyshort 
reviewer @tomblench

### CODE REVIEW

**Comment by rhyshort** [comment]
General comment, new files need the licence header

**Comment by rhyshort** [comment]
+1

File: Classes/common/Utils/CDTChangedArray.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  - addObject:
+  - removeLastObject
+  - replaceObjectAtIndex:withObject:
+
+ The changed flag will only be set for modifications through this class, not
+ when the underlying array changes itself, therefore it's important
+ not to let the contained array "escape" from this class if all changes
+ need to be tracked.
+ */
+@interface CDTChangedArray : NSMutableArray <CDTChangedObserver>
```
Comment (tomblench) [inline Classes/common/Utils/CDTChangedArray.h]: Why does the array itself have to adopt this protocol? The interested party will set the appropriate delegate object and if they don't want a delegate they can access `isChanged` directly.

File: Classes/common/Utils/CDTChangedArray.h [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ Init with array. This constructor must be used.
+ */
+- (instancetype)initWithMutableArray:(nonnull NSMutableArray *)array;
+
+/**
+ Set to YES if this dictionary has been modified.
+ */
+@property (nonatomic) bool isChanged;
```
Comment (tomblench) [inline Classes/common/Utils/CDTChangedArray.h]: `isChanged` sounds awkward although `isXXX` is standard for booleans. Does `hasChanged` sound too much like a delegate callback? Or `wasChanged`, or `didChange`?

File: Tests/Tests/Utils/CDTChangedDictionaryJSONWrappingTests.m [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+#import <CDTChangedDictionary.h>
+#import <CDTChangedArray.h>
+
+@interface CDTChangedDictionaryJSONWrappingTests : XCTestCase
+
+@property (nonatomic, strong) CDTChangedDictionary *dictionary;
+
+@end
+
+@implementation CDTChangedDictionaryJSONWrappingTests
```
Comment (tomblench) [inline Tests/Tests/Utils/CDTChangedDictionaryJSONWrappingTests.m]: would it be useful to test the `changed` true → false transition which occurs when calling `dictionaryWrappingContents`?

File: Classes/common/Utils/CDTChangedArray.h (hunk 1 above)
Comment (mikerhodes) [inline Classes/common/Utils/CDTChangedArray.h]: The array and dictionary are delegates of the objects they contain.

File: Classes/common/Utils/CDTChangedArray.h (hunk 2 above)
Comment (mikerhodes) [inline Classes/common/Utils/CDTChangedArray.h]: Hmm. I think `isChanged` is standard; one is asking about the current state?

File: Tests/Tests/Utils/CDTChangedDictionaryJSONWrappingTests.m (hunk 3 above)
Comment (mikerhodes) [inline Tests/Tests/Utils/CDTChangedDictionaryJSONWrappingTests.m]: That should never be externally visible -- the newly created dictionary doesn't leave the `dictionaryWrappingContents` scope before `changed` is set to `true`.

File: Tests/Tests/Utils/CDTChangedDictionaryJSONWrappingTests.m [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+    NSDictionary *dictionary = @{
+        @"dict" :
+            @{@"array" : @[ @"one", @"two" ], @"dict" : @{@"one" : @"two"}, @"hello" : @"world"},
+
+        @"array" : @[ @[ @"foo", @YES ], @{@"foo" : @YES}, @"two" ]
+    };
+
+    self.dictionary = [CDTChangedDictionary dictionaryWrappingContents:dictionary];
+}
```
Comment (tomblench) [inline Tests/Tests/Utils/CDTChangedDictionaryJSONWrappingTests.m]: Would it be useful to have `XCTAssertFalse(self.dictionary.isChanged);` either at the end of setup or at the start of each test?

Comment (mikerhodes) [inline Tests/Tests/Utils/CDTChangedDictionaryJSONWrappingTests.m]: I test this here https://github.com/cloudant/CDTDatastore/pull/189/files#diff-7de81db5bbf96ef38818303145945047R44

**Comment by tomblench** [comment]
OK +1

**Comment by mikerhodes** [comment]
Squashing and merging.

**Pull request merged**